### ISSUE TITLE
Pytorch example on how to implement on existing project

### ISSUE THREAD

**Issue by marisancans** [NVIDIA/DALI#2661]
Hello, is there any example on how can I transfer from pytorch dataset and default dataloader to DALI? It seems that its a lot easier to just start over rather than trying to figure out how to implement this. Given examples for pytorch are too basic in my opinion.

By this I mean that I have custom Dataset class in pytorch and I cant figure out how to transfer to DALI. When I look for tutorials its either: pytorch lighting or uses some neatly ready solutions like imgenet MXNet and Caffee.

Thankis

**Comment by JanuszL** [comment]
Hi,
I would not focus on how to transfer pipeline 1:1 from PyTorch but rather think about what you want to do and express it with available DALI operators.
The examples we provide are [how to build the pipeline using DALI ](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/index.html)and then [how to integrate if with PyTorch](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/plugins/pytorch_tutorials.html).
Could you tell us what kind of processing flow you want to implement?

**Comment by marisancans** [comment]
hey, had forgotten with this for a while and Im trying to implement DALI again. However it is more of a fight for me to do it unfortunately. 

One of the biggest problems I have is figuring out how to load the data. All of the examples I have seen assumes that the format is like this: Have a batch size of N and then load N images, all simple, all good
Example:
batch = [ image_1, image_2, image_3, image_4]

However, in my case I have such format N x M, where M is also varies
Example:
batch = [[image_1, image_2], [image_1, image_2, image_3, image_4], [image_1, image_2, image_3]]

How do I acomplish this? Thanks

**Comment by JanuszL** [comment]
Hi @marisancans,

As I understand you want to load a sequence of the images, where each sample may have one or more stacked images, right?

**Comment by marisancans** [comment]
That was a fast response!
I will try to visualize it more so we are on the same page
![Untitled Diagram](https://user-images.githubusercontent.com/16538903/114034683-14172480-9887-11eb-94be-851b64000858.png)

**Comment by JanuszL** [comment]
@marisancans,

The first thing is not possible but 
![image](https://user-images.githubusercontent.com/39967756/114036141-1aee6900-9880-11eb-9d5e-8635eb5d9e86.png) should be doable.
DALI supports variable batch size iteration to iteration. To use it please check [the `external_source` operator](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/supported_ops.html#nvidia.dali.fn.external_source). [This example](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/general/data_loading/external_input.html) should guide you.
To achieve the desired result each iteration just return the number of samples you want. Make sure to set `batch_size` in the pipeline to the maximum number of samples you want to feed into processing. You may also want to adjust the `parallel` and `prefetch_queue_depth` options of the `external_source` operator to get the best performance.

**Comment by marisancans** [comment]
I think I got your idea, but I got an error I cant figure out now: 
`Assert on "this->IsDenseTensor()" failed: All tensors in the input TensorList must have the same shape and be densely packed.`

Heres the iterator idea you talked about, it returns a tuple where the elements in that tuple varies 0 to 30

```
from torch.utils.data import IterableDataset
import numpy as np
from helpers.fs import ls_files

class ExternalInputIterator(object):
    def __init__(self, batch_size):
        self.batch_size = batch_size

    def __iter__(self):
        self.i = 0
        return self

    def __next__(self):
        images = []
        labels = []
        image_paths_ord = []

        sample_path_1 = "/media/maris/data/vbfer_v2_preprocessor_datasets/crema_d/emotions/anger/0/detected_faces"
        sample_path_2 = "/media/maris/data/vbfer_v2_preprocessor_datasets/crema_d/emotions/happiness/67/detected_faces"
        sample_path_3 = "/media/maris/data/vbfer_v2_preprocessor_datasets/crema_d/emotions/happiness/1330/detected_faces"

        sample_paths = [sample_path_1, sample_path_2, sample_path_3]
        sample_path = sample_paths[self.i]
        image_paths = ls_files(sample_path)

        print(sample_path)

        # Here image_paths can be any size 1-30
        for image_path in image_paths:

            f = open(image_paths[0], "rb")
            image = np.frombuffer(f.read(), dtype=np.uint8) # don't decode the image just read the bytes!
            images.append(image)

            label = np.array([1])  # some label
            labels.append(label)

            image_path_ord = [ord(x) for x in image_path] #this is a hacky trick used to pass image_paths(strings) through DALI.
            image_path_ord = np.array([1, 2, 3])
            # image_path_ord = np.array(image_path_ord, dtype=np.int32)
            image_paths_ord.append(image_path_ord)

        print("SELF I:", self.i)
        if self.i == 2:
            raise StopIteration
        
        self.i += 1
        print(len(images), len(labels), len(image_paths_ord))

        print('lens Images', [len(x) for x in images])
        print('lens labels', [len(x) for x in labels])
        print('lens image_paths_ord', [len(x) for x in image_paths_ord])

        images = np.ascontiguousarray(images)
        labels = np.ascontiguousarray(labels)
        image_paths_ord = np.ascontiguousarray(image_paths_ord)

        return (images, labels, image_paths_ord) # This always returns the same data for testing
```


Here is a pipeline
```
from nvidia.dali.pipeline import Pipeline
import nvidia.dali.ops as ops
import nvidia.dali.types as types

#Using ExternalSource
class SimplePipeline(Pipeline):
    #Define the operations in the pipeline
    def __init__(self, external_datasource, batch_size=16, num_threads=2, device_id=0, resolution=256, crop=224, is_train=True):
        super(SimplePipeline, self).__init__(batch_size, num_threads, device_id, seed=12)
        # Define Input nodes
        self.jpegs = ops.ExternalSource()
        self.in_labels = ops.ExternalSource()
        self.in_paths = ops.ExternalSource()
        ## Or pass source straight to ExternalSource this way you won't have do iter_setup.
        # self.jpegs,self.labels,self.paths=ops.ExternalSource(source=self.make_batch, num_outputs=3)
        
        # Define ops
        self.decode = ops.ImageDecoder(device="mixed", output_type=types.RGB)
        self.res = ops.Resize(device="gpu", resize_x=resolution,resize_y=resolution)
        self.normalize= ops.CropMirrorNormalize(device="gpu",
                                            dtype=types.FLOAT,
                                            output_layout=types.NCHW,
                                            crop=(crop, crop),
                                            mean=[0.485 * 255,0.456 * 255,0.406 * 255],
                                            std=[0.229 * 255,0.224 * 255,0.225 * 255])
                                            
        self.path_pad=ops.Pad(fill_value=ord("?"),axes = (0,)) # We need to pad image_paths because we need the shapes to match.need dense tensor

        self.iterator = iter(external_datasource)

    # The external source should be fed batches
    # I prefer to batch-ify things here because it keeps things compatible with an IterableDataset
    def make_batch(self):
        imgs = []
        labels = []
        paths = []
        for _ in range(self.batch_size):
            i, l, p = next(self.iterator)
            imgs.append(i)
            labels.append(l)
            paths.append(p)

        return (imgs, labels, paths)
    
    # How the operations in the pipeline are used
    # Connect your input nodes to your ops
    def define_graph(self):
        
        self.images = self.jpegs()
        self.labels = self.in_labels()
        self.paths = self.in_paths()
        images = self.decode(self.images)
        images = self.res(images)
        images = self.normalize(images)
            
        paths = self.path_pad(self.paths)
    
        return (images, self.labels, paths)

    # Only needed when using ExternalSource
    # Connect the dataset outputs to external Sources
    def iter_setup(self):
        (images, labels, paths) = next(self.iterator)

        self.feed_input(self.images, images)
        self.feed_input(self.labels, labels)
        self.feed_input(self.paths, paths)
```

This is how I create the pipeline iterator:
```

def make_pipeline(dataset, args, device_index=0, return_keys=["images", "labels","image_path"], num_threads=2, is_train=True):

    # dataset here is ExternalInputIterator
    pipeline = SimplePipeline(dataset, batch_size=args["batch_size"], num_threads=num_threads,
                        device_id=device_index, resolution=args["resolution"], crop=args["crop"], is_train=is_train)

    pipeline_iterator = DALIGenericIterator(pipeline, return_keys)
    return pipeline_iterator
```


And this is how I iterate
```
 for n_batch, batch in enumerate(data_loader):
                    print(len(batch[0]['image_path']))
                    continue
```

I cant figure out why the error complains about different sizes
Sorry to put a block of code out here like this, really wanna try out the DALI

**Comment by marisancans** [comment]
It seems im doing something similar to this: https://github.com/NVIDIA/DALI/issues/2007#issue-634520448

**Comment by marisancans** [comment]
The first Iteration works, error pops up on the next iteration

**Comment by JanuszL** [comment]
I see that samples in the batch should be uniform as you apply to resize and pad. The only thing that may be missing is `dynamic_shape` in the DALIGenericIterator. If you enable it allows iteration to iteration variability in the shape returned from the DALIGenericIterator.

**Comment by marisancans** [comment]
Thanks, that did the trick later on, the error was caused by padding operation
`paths = self.path_pad(self.paths)`

Im not sure why, because all data that I tried to pad were in the same lengths, but I think its working now
